ligo/src/proposal.mligo

```diff
@@ -4,21 +4,361 @@
+  let proposal_key = vote_param.proposal_key in
+  let proposal =
+    match Map.find_opt proposal_key store.proposals with
+      None -> (failwith("PROPOSAL_NOT_EXIST") : proposal)
+    | Some p -> p
+    in
+
+  let store = freeze (vote_param.vote_amount, author, store) in
+
+  let proposal =
+    if vote_param.vote_type
+      then { proposal with upvotes = proposal.upvotes + vote_param.vote_amount }
+      else { proposal with downvotes = proposal.downvotes + vote_param.vote_amount }
+    in
+  let proposal =
+    { proposal with
+      voters = (author, vote_param.vote_amount) :: proposal.voters
+    } in
+
+  { store with
+    proposals = Map.add proposal_key proposal store.proposals
+  }
+
+[@inline]
+let check_vote_limit_reached
+    (config, proposal, vote_param : config * proposal * vote_param): unit =
+  if config.max_votes < proposal.upvotes + proposal.downvotes + vote_param.vote_amount
+    then failwith("MAX_VOTES_REACHED")
+    else ()
+
+let vote(votes, config, store : vote_param_permited list * config * storage): return =
+  let store = ensure_not_migrated store in
+  let accept_vote = fun (store, pp : storage * vote_param_permited) ->
+    let (param, author, store) = verify_permit_protected_vote (pp, store) in
+    let proposal = check_if_proposal_exist (param.proposal_key, store) in
+    let checked = check_vote_limit_reached (config, proposal, param) in
+    let checked = ensure_voting_period_is_not_over (proposal, store) in
+    let checked = check_voter_unfrozen_token (param, author, store) in
+    let store = submit_vote (param, author, store) in
+    store
+    in
+  ( ([] : operation list)
+  , List.fold accept_vote votes store
+  )
+
+// -----------------------------------------------------------------
+// Admin entrypoints
+// -----------------------------------------------------------------
+
+// Update voting period of all ongoing and new proposals.
+[@inline]
+let set_voting_period(new_period, config, store : voting_period * config * storage): return =
+  let store = ensure_not_migrated store in
+  let store = authorize_admin store in
+  let checked =
+    if   config.max_voting_period < new_period
+      || config.min_voting_period > new_period
+      then failwith("OUT_OF_BOUND_VOTING_PERIOD")
+      else ()
+    in
+  let store = { store with voting_period = new_period } in
+  (([] : operation list), store)
+
+// Update quroum_threshold. The new quorum_threshold affects
+// all ongoing and new proposals.
+[@inline]
+let set_quorum_threshold(new_threshold, config, store : quorum_threshold * config * storage): return =
+  let store = ensure_not_migrated store in
+  let store = authorize_admin store in
+  let checked =
+    if   config.max_quorum_threshold < new_threshold
+      || config.min_quorum_threshold > new_threshold
+      then failwith("OUT_OF_BOUND_QUORUM_THRESHOLD")
+      else ()
+    in
+  let store = { store with quorum_threshold = new_threshold } in
+  (([] : operation list), store)
```

Comment:
```suggestion
  let store =
    if config.max_voting_period < new_period || config.min_voting_period > new_period
    then (failwith("OUT_OF_BOUND_VOTING_PERIOD") : storage)
    else { store with voting_period = new_period }
  in (([] : operation list), store)
```
I just think it's a bit easier to read

Reply:
Mm, I tried this and the condition does not fit into 80 chars now :confused: 

While in other places like function signatures this is not crucial, here it seems to be.

Reply:
What do you mean it's crucial? For compilation?

If for reading it's ok, the suggestion was more about returning a `storage` for the `if` by updating it in the successful branch instead of afterwards.

Reply:
Crucial for readability, e.g. when I have two panes opened then part of the condition does not fit into the screen.

Though the second part - yep, will do now.

Reply:
Ok, I agree. For a moment there I thought you had found another oddity of the cameligo syntax :sweat_smile: